### PR TITLE
New Agent Abstraction [Part 1 of Multi-Agent RL]

### PR DESCRIPTION
**Pull request by ASzot**
## Motivation and Context

Refactored trainer code to interface with a new agent abstraction which wraps the rollouts, policy, and updaters. This will enable multi-agent RL in Part 2 of the Multi-Agent RL PR. 

## How Has This Been Tested

Integration tests with training policies. 

## Types of changes

- Docs change / refactoring / dependency upgrade

### CODE REVIEW

File: habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py [hunk 1]
```diff
@@ -315,69 +225,43 @@ def _init_train(self, resume_state=None):
         ):
             os.makedirs(self.config.habitat_baselines.checkpoint_folder)
 
-        self._setup_actor_critic_agent(ppo_cfg)
-        if resume_state is not None:
-            self.agent.load_state_dict(resume_state["state_dict"])
-            self.agent.optimizer.load_state_dict(resume_state["optim_state"])
-        if self._is_distributed:
-            self.agent.init_distributed(find_unused_params=False)  # type: ignore
+        logger.add_filehandler(self.config.habitat_baselines.log_file)
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py:228]: Is this the best place to put this logger change? I would expect it to be better at the very beginning. Like in run.py.

File: habitat-baselines/habitat_baselines/rl/ppo/trainer_agent.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        self._percent_done_fn = percent_done_fn
+
+        self._lr_scheduler = LambdaLR(
+            optimizer=self._updater.optimizer,
+            lr_lambda=lambda x: linear_lr_schedule(
+                percent_done=percent_done_fn()
+            ),
+        )
+
+    def post_init(self, create_rollouts_fn: Optional[Callable] = None) -> None:
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/trainer_agent.py:95]: I do not like the name `post_init` or the docstring `Called after the constructor.`
It makes it sound like this is an automatic callback, but it is not, it is called manually by the trainer, which increases code entanglement. 
Is there a reason this cannot be called in __init__ ?
If so, can we call this method by what it does "initialize_rollout" and make sure that the error message in case this method is not called before using the trainer_agent the error message is easy to interpret?

File: habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py [hunk 3]
```diff
@@ -398,11 +282,11 @@ def save_checkpoint(
             None
         """
         checkpoint = {
-            "state_dict": self.agent.state_dict(),
+            **self._agent.get_save_state(),
             "config": self.config,
         }
         if extra_state is not None:
-            checkpoint["extra_state"] = extra_state
+            checkpoint["extra_state"] = extra_state  # type: ignore
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py:289]: Why the added `# type: ignore` did the code change or was this broken to begin with ?

File: habitat-baselines/habitat_baselines/rl/ppo/trainer_agent.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        Called after the updater has called `update` and the rollout `after_update` is called.
+        """
+
+        if self._ppo_cfg.use_linear_lr_decay:
+            self._lr_scheduler.step()  # type: ignore
+
+    def pre_rollout(self):
+        """
+        Called before a rollout is collected.
+        """
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/trainer_agent.py:272]: Same comment as above, this is manually called but sounds like it is a callback.
I would call this `update_clip_parameter` or even better call this automatically.

File: habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py [hunk 5]
```diff
@@ -937,14 +808,11 @@ def _eval_checkpoint(
 
         self._init_envs(config, is_eval=True)
 
-        self._setup_actor_critic_agent(ppo_cfg)
         action_shape, discrete_actions = get_action_space_info(
-            self.policy_action_space
+            self._agent.policy_action_space
         )
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py:813]: Can we use the new EnvironmentSpec here instead of accessing the agent?

File: habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py [hunk 6]
```diff
@@ -1038,13 +907,13 @@ def _eval_checkpoint(
             # NB: Move actions to CPU.  If CUDA tensors are
             # sent in to env.step(), that will create CUDA contexts
             # in the subprocesses.
-            if is_continuous_action_space(self.env_action_space):
+            if is_continuous_action_space(self._env_spec.action_space):
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py:910]: nit, make is_continuous_action_space a method on EnvironmentSpec so that the `is_continuous_action_space` is not isolated.

File: habitat-baselines/habitat_baselines/rl/ppo/trainer_agent.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        self._actor_critic = self._create_policy()
+        self._actor_critic.to(self._device)
+        self._policy_action_space = self._actor_critic.get_policy_action_space(
+            env_spec.action_space
+        )
+        self._updater = self._create_updater(self._actor_critic)
+        if resume_state is not None:
+            self._updater.load_state_dict(resume_state["state_dict"])
+            self._updater.optimizer.load_state_dict(
+                resume_state["optim_state"]
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/trainer_agent.py:75]: Shouldn't the learning rate also be loaded here?

File: habitat-baselines/habitat_baselines/rl/ppo/trainer_agent.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+            )
+
+        logger.info(
+            "agent number of parameters: {}".format(
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/trainer_agent.py:79]:
```suggestion
            "Agent number of parameters: {}".format(
```
nit, is there any other information we want to give the user here ? Like the observation/action space ?

File: habitat-baselines/habitat_baselines/rl/ppo/trainer_agent.py [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+                self._config.habitat_baselines.distrib_updater_name
+            )
+        else:
+            updater_cls = baseline_registry.get_updater(
+                self._config.habitat_baselines.updater_name
+            )
+        return updater_cls.from_config(actor_critic, self._ppo_cfg)
+
+    @property
+    def policy_action_space(self):
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/trainer_agent.py:129]: I would replace this with environment_spec here. Let's make use of it.

File: habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py [hunk 10]
```diff
@@ -567,29 +452,34 @@ def _collect_rollout_step(self):
...
+            self._ppo_cfg.use_gae,
+            self._ppo_cfg.gamma,
+            self._ppo_cfg.tau,
         )
 
-        self.agent.train()
+        self._agent.train()
+
+        losses = self._agent.updater.update(self._agent.rollouts)
+        self._agent.rollouts.after_update()
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py:479]: Is it simpler if self._agent.after_update() calls self._agent.rollouts.after_update() automatically ?

File: habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py [hunk 11]
```diff
@@ -567,29 +452,34 @@ def _collect_rollout_step(self):
...
+            next_value,
+            self._ppo_cfg.use_gae,
+            self._ppo_cfg.gamma,
+            self._ppo_cfg.tau,
         )
 
-        self.agent.train()
+        self._agent.train()
+
+        losses = self._agent.updater.update(self._agent.rollouts)
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py:478]: I do not like calling directly rollouts and updater on _agent because it seems like there is a specific order these methods need to be called in. I would push this complexity down and create an `agent.update()` that would call 
```
self.rollouts.compute_returns(
            next_value,
            self._ppo_cfg.use_gae,
            self._ppo_cfg.gamma,
            self._ppo_cfg.tau,
        )
self.train()
self.updater.update(self._agent.rollouts)
self.rollouts.after_update()
self.after_update()
self.eval()
```
and maybe even the block inside with `inference_mode()`

If we do not want to do this (not sure why) we should at least isolate this logic into a static method that takes as input the agent and the _ppo_cfg paraeters.
